### DISCUSSION TITLE
Update event is not received when a user is removed from a record.

### DISCUSSION THREAD

**Discussion by TimeLex** [pocketbase/pocketbase#5127] [Q&A] [answered]
Hi, I'm new to PocketBase and databases in general. So I'm trying to build a basic chat app to understand the basics.

I have a chats collection. Each chat record stores all users that can write to the chat and each user can only see chats they are in (list/search rule and view rule: `users.id ?= @request.auth.id`). To recognize when users are added to or kicked out of a chat, the real-time events are subscribed to. And this all works fine, except in one case:
When a user is kicked out of a chat, it does not fulfill the read rules of the chat, so the user does not get the update that he has been removed.

I tried changing the reading rules, but every change that fixed the problem also resulted in every user seeing all chats, which I don't want. I also tried a bit with hooks, but couldn't get it to work.

By the way, I use Dart, so it's preferred, but I can also understand JavaScript code.

Thanks in advance!

**Comment by ganigeorgiev** [comment]
We don't currently send events for records that don't satisfy the List/View API rule of the related collection.

It exists in the roadmap to consider it at some point in time to track the "unavaiable" state but remains a low priority (see https://github.com/pocketbase/pocketbase/issues/4717, https://github.com/pocketbase/pocketbase/discussions/4592, https://github.com/pocketbase/pocketbase/issues/2522).

**Reply by ganigeorgiev** [reply, marked as answer]
> I also tried a bit with hooks, but couldn't get it to work.

What did you try? With the JS `pb_hooks` it will be a little more complicated and less performant but it should be doable. If we assume that the "chat" racord contains the related users as multiple `relation` field and you just want a "kicked" event you can send it manually by using something like (_I haven't tested it_):

```js
onModelAfterUpdate((e) => {
    const chat         = e.model
    const oldUsers     = chat.originalCopy().getStringSlice("users")
    const newUsers     = chat.getStringSlice("users")
    const removedUsers = oldUsers.filter((id) => !newUsers.includes(id));

    if (!removedUsers.length) {
        return // no removed users
    }

    const clients = $app.subscriptionsBroker().clients();

    for (const clientId in clients) {
        const client = clients[clientId];
        const user   = client.get("authRecord");
  
        // not an authenticated user or not in the list of the removed ones
        if (!user || !removedUsers.includes(user.id) || !client.hasSubscription("kicked")) {
            continue
        }

        client.send(new SubscriptionMessage({
            name: "kicked",
            data: JSON.stringify({ chatId: chat.id })
        }))
    }
}, "chats")
```

Then on the client-side, you can subscribe to the custom event using something like:
```dart
pb.realtime.subscribe("kicked", callback)
```

_edit: added optional `client.hasSubscription` check to avoid sending the event if not explicitly subscribed to._

**Reply by TimeLex** [reply]
This works perfectly. Thanks!
